File: flutter/sherpa_onnx/lib/src/offline_speech_denoiser.dart
// Copyright (c)  2025  Xiaomi Corporation
import 'dart:ffi';
import 'dart:typed_data';

import 'package:ffi/ffi.dart';
import './sherpa_onnx_bindings.dart';
import './offline_speech_denoiser_config.dart';

export './offline_speech_denoiser_config.dart';

/// Offline speech denoiser.
class OfflineSpeechDenoiser {
  OfflineSpeechDenoiser.fromPtr({required this.ptr, required this.config});

  OfflineSpeechDenoiser._({required this.ptr, required this.config});

  /// Create an offline denoiser from [config].
  factory OfflineSpeechDenoiser(OfflineSpeechDenoiserConfig config) {
    if (SherpaOnnxBindings.sherpaOnnxCreateOfflineSpeechDenoiser == null) {
      throw Exception("Please initialize sherpa-onnx first");
    }

    final c = calloc<SherpaOnnxOfflineSpeechDenoiserConfig>();
    c.ref.model.gtcrn.model = config.model.gtcrn.model.toNativeUtf8();
    c.ref.model.dpdfnet.model = config.model.dpdfnet.model.toNativeUtf8();
    c.ref.model.dpdfnet.attenuationLimitDb =
        config.model.dpdfnet.attenuationLimitDb;

    c.ref.model.numThreads = config.model.numThreads;
    c.ref.model.debug = config.model.debug ? 1 : 0;
    c.ref.model.provider = config.model.provider.toNativeUtf8();

    final ptr =
        SherpaOnnxBindings.sherpaOnnxCreateOfflineSpeechDenoiser?.call(c) ??
            nullptr;

    calloc.free(c.ref.model.provider);
    calloc.free(c.ref.model.gtcrn.model);
    calloc.free(c.ref.model.dpdfnet.model);
    calloc.free(c);

    if (ptr == nullptr) {
      throw Exception(
          "Failed to create offline speech denoiser. Please check your config");
    }

    return OfflineSpeechDenoiser._(ptr: ptr, config: config);
  }

  /// Release the native denoiser.
  void free() {
    if (SherpaOnnxBindings.sherpaOnnxDestroyOfflineSpeechDenoiser == null) {
      throw Exception("Please initialize sherpa-onnx first");
    }

    if (ptr == nullptr) {
      return;
    }

    SherpaOnnxBindings.sherpaOnnxDestroyOfflineSpeechDenoiser?.call(ptr);
    ptr = nullptr;
  }

  /// Denoise one chunk or a complete waveform.
  DenoisedAudio run({required Float32List samples, required int sampleRate}) {
    if (SherpaOnnxBindings.sherpaOnnxOfflineSpeechDenoiserRun == null) {
      throw Exception("Please initialize sherpa-onnx first");
    }

    if (ptr == nullptr) {
      return DenoisedAudio(samples: Float32List(0), sampleRate: 0);
    }

    final n = samples.length;
    final Pointer<Float> psamples = calloc<Float>(n);

    final pList = psamples.asTypedList(n);
    pList.setAll(0, samples);

    final p = SherpaOnnxBindings.sherpaOnnxOfflineSpeechDenoiserRun
            ?.call(ptr, psamples, n, sampleRate) ??
        nullptr;

    calloc.free(psamples);

    if (p == nullptr) {
      return DenoisedAudio(samples: Float32List(0), sampleRate: 0);
    }

    final sampleRateOut = p.ref.sampleRate;
    final nOut = p.ref.n;
    Float32List newSamples = Float32List(0);
    if (nOut > 0 && p.ref.samples != nullptr) {
      newSamples = Float32List.fromList(p.ref.samples.asTypedList(nOut));
    }

    SherpaOnnxBindings.sherpaOnnxDestroyDenoisedAudio?.call(p);

    return DenoisedAudio(samples: newSamples, sampleRate: sampleRateOut);
  }

  /// Return the expected sample rate for this denoiser.
  int get sampleRate {
    if (SherpaOnnxBindings.sherpaOnnxOfflineSpeechDenoiserGetSampleRate ==
        null) {
      throw Exception("Please initialize sherpa-onnx first");
    }

    if (ptr == nullptr) {
      return 0;
    }

    return SherpaOnnxBindings.sherpaOnnxOfflineSpeechDenoiserGetSampleRate
            ?.call(ptr) ??
        0;
  }

  Pointer<SherpaOnnxOfflineSpeechDenoiser> ptr;
  OfflineSpeechDenoiserConfig config;
}

File: sherpa-onnx/rust/sherpa-onnx/src/offline_speech_denoiser.rs
//! Offline speech denoising.
//!
//! Supported model families mirror the native API and currently include GTCRN
//! and DPDFNet. See the repository examples:
//!
//! - `rust-api-examples/examples/offline_speech_enhancement_gtcrn.rs`
//! - `rust-api-examples/examples/offline_speech_enhancement_dpdfnet.rs`

use crate::utils::to_c_ptr;
use sherpa_onnx_sys as sys;
use std::ffi::CString;
use std::ptr;
use std::slice;

#[derive(Clone, Debug, Default)]
/// GTCRN model path for offline denoising.
pub struct OfflineSpeechDenoiserGtcrnModelConfig {
    pub model: Option<String>,
}

impl OfflineSpeechDenoiserGtcrnModelConfig {
    pub(crate) fn to_sys(
        &self,
        cstrings: &mut Vec<CString>,
    ) -> sys::OfflineSpeechDenoiserGtcrnModelConfig {
        sys::OfflineSpeechDenoiserGtcrnModelConfig {
            model: to_c_ptr(&self.model, cstrings),
        }
    }
}

#[derive(Clone, Debug, Default)]
/// DPDFNet model path for offline denoising.
pub struct OfflineSpeechDenoiserDpdfNetModelConfig {
    pub model: Option<String>,
}

impl OfflineSpeechDenoiserDpdfNetModelConfig {
    pub(crate) fn to_sys(
        &self,
        cstrings: &mut Vec<CString>,
    ) -> sys::OfflineSpeechDenoiserDpdfNetModelConfig {
        sys::OfflineSpeechDenoiserDpdfNetModelConfig {
            model: to_c_ptr(&self.model, cstrings),
        }
    }
}

#[derive(Clone, Debug)]
/// Aggregate model configuration for [`OfflineSpeechDenoiser`].
///
/// Configure exactly one model family in normal use.
pub struct OfflineSpeechDenoiserModelConfig {
    pub gtcrn: OfflineSpeechDenoiserGtcrnModelConfig,
    pub dpdfnet: OfflineSpeechDenoiserDpdfNetModelConfig,
    pub num_threads: i32,
    pub debug: bool,
    pub provider: Option<String>,
}

impl Default for OfflineSpeechDenoiserModelConfig {
    fn default() -> Self {
        Self {
            gtcrn: Default::default(),
            dpdfnet: Default::default(),
            num_threads: 1,
            debug: false,
            provider: Some("cpu".to_string()),
        }
    }
}

impl OfflineSpeechDenoiserModelConfig {
    pub(crate) fn to_sys(
        &self,
        cstrings: &mut Vec<CString>,
    ) -> sys::OfflineSpeechDenoiserModelConfig {
        sys::OfflineSpeechDenoiserModelConfig {
            gtcrn: self
                .gtcrn
                .to_sys(cstrings),
            num_threads: self.num_threads,
            debug: self.debug as i32,
            provider: to_c_ptr(&self.provider, cstrings),
            dpdfnet: self
                .dpdfnet
                .to_sys(cstrings),
        }
    }
}

#[derive(Clone, Debug, Default)]
/// Denoised samples returned from an offline or online denoiser.
pub struct DenoisedAudio {
    pub samples: Vec<f32>,
    pub sample_rate: i32,
}

impl DenoisedAudio {
    pub(crate) fn from_ptr(ptr: *const sys::DenoisedAudio) -> Self {
        if ptr.is_null() {
            return Self::default();
        }

        unsafe {
            let n = (*ptr)
                .n
                .max(0) as usize;
            let samples = if (*ptr)
                .samples
                .is_null()
                || n == 0
            {
                vec![]
            } else {
                slice::from_raw_parts((*ptr).samples, n).to_vec()
            };
            let sample_rate = (*ptr).sample_rate;
            sys::SherpaOnnxDestroyDenoisedAudio(ptr);
            Self {
                samples,
                sample_rate,
            }
        }
    }
}

#[derive(Clone, Debug, Default)]
/// Top-level configuration for [`OfflineSpeechDenoiser`].
pub struct OfflineSpeechDenoiserConfig {
    pub model: OfflineSpeechDenoiserModelConfig,
}

impl OfflineSpeechDenoiserConfig {
    fn to_sys(&self, cstrings: &mut Vec<CString>) -> sys::OfflineSpeechDenoiserConfig {
        sys::OfflineSpeechDenoiserConfig {
            model: self
                .model
                .to_sys(cstrings),
        }
    }
}

/// Offline speech denoiser.
pub struct OfflineSpeechDenoiser {
    ptr: *const sys::OfflineSpeechDenoiser,
}

// SAFETY: The sherpa-onnx C library is thread-safe for single-object usage.
unsafe impl Send for OfflineSpeechDenoiser {}
unsafe impl Sync for OfflineSpeechDenoiser {}

impl OfflineSpeechDenoiser {
    /// Create a denoiser from `config`.
    pub fn create(config: &OfflineSpeechDenoiserConfig) -> Option<Self> {
        let mut cstrings = Vec::new();
        let sys_config = config.to_sys(&mut cstrings);
        let ptr = unsafe { sys::SherpaOnnxCreateOfflineSpeechDenoiser(&sys_config) };
        if ptr.is_null() {
            None
        } else {
            Some(Self { ptr })
        }
    }

    /// Denoise one chunk or a complete waveform.
    pub fn run(&self, samples: &[f32], sample_rate: i32) -> DenoisedAudio {
        let samples_ptr = if samples.is_empty() {
            ptr::null()
        } else {
            samples.as_ptr()
        };
        let ptr = unsafe {
            sys::SherpaOnnxOfflineSpeechDenoiserRun(
                self.ptr,
                samples_ptr,
                samples.len() as i32,
                sample_rate,
            )
        };
        DenoisedAudio::from_ptr(ptr)
    }

    /// Return the model sample rate expected by this denoiser.
    pub fn sample_rate(&self) -> i32 {
        unsafe { sys::SherpaOnnxOfflineSpeechDenoiserGetSampleRate(self.ptr) }
    }
}

impl Drop for OfflineSpeechDenoiser {
    fn drop(&mut self) {
        unsafe { sys::SherpaOnnxDestroyOfflineSpeechDenoiser(self.ptr) }
    }
}
